wallet: skip R-value signature grinding for external signers

Sjors opened the pull request:
When producing a dummy signature for the purpose of estimating the transaction fee, do not assume an external signer performs R-value grinding on the signature.

In particular, this avoids a scenario where the fee rate is 1 sat / vbyte and a transaction with a 72 byte signature is not accepted into our mempool.

Suggested testing:
1. On master, launch with `-signet` and create an external signer wallet using e.g. a Trezor and HWI, see [guide](https://github.com/bitcoin/bitcoin/blob/master/doc/external-signer.md#example-usage) (with the GUI it should "just work" once you have the HWI path configured).
2. Create a few addresses and fund them from the faucet: https://signet.bc-2.jp/ (wait for confirmation)
3. Create another address, and now send the entire wallet to it, set the fee to 1 sat/byte
4. Most likely this transaction never gets broadcast and you won't see it on the [signet explorer](https://explorer.bc-2.jp) 

5. With this PR, try again.
6. Check the explorer and inspect the transaction. Each input witness starts with either `30440220` (R has 32 bytes) or `30440221` (R has 33 bytes). See this explainer for [DER encoding](https://bitcoin.stackexchange.com/questions/92680/what-are-the-der-signature-and-sec-format).

Fixes #26030

Sjors committed:
wallet: annotate bools in descriptor SPKM FillPSBT()

Sjors commented:
Added a release note.

luke-jr commented:
A functional test would be nice

luke-jr commented on src/psbt.cpp:
```diff
@@ -320,7 +320,7 @@ PrecomputedTransactionData PrecomputePSBTData(const PartiallySignedTransaction&
 }
 
-bool SignPSBTInput(const SigningProvider& provider, PartiallySignedTransaction& psbt, int index, const PrecomputedTransactionData* txdata, int sighash,  SignatureData* out_sigdata, bool finalize)
+bool SignPSBTInput(const SigningProvider& provider, PartiallySignedTransaction& psbt, int index, const PrecomputedTransactionData* txdata, int sighash,  SignatureData* out_sigdata, bool finalize, bool grind)
```
nit: `grind` is a bit unclear, maybe `require_optimal_signature_size` or `!use_max_sig` (like in wallet_tests)?

luke-jr commented on src/wallet/scriptpubkeyman.cpp:
```diff
@@ -618,7 +618,7 @@ SigningResult LegacyScriptPubKeyMan::SignMessage(const std::string& message, con
 }
 
-TransactionError LegacyScriptPubKeyMan::FillPSBT(PartiallySignedTransaction& psbtx, const PrecomputedTransactionData& txdata, int sighash_type, bool sign, bool bip32derivs, int* n_signed, bool finalize) const
+TransactionError LegacyScriptPubKeyMan::FillPSBT(PartiallySignedTransaction& psbtx, const PrecomputedTransactionData& txdata, int sighash_type, bool sign, bool bip32derivs, int* n_signed, bool finalize, bool grind) const
```
`grind` gets ignored here. What if a watch-only input is used, and the signer for it can't grind?

luke-jr reviewed (commented):
I wonder if it would make sense to move `grind` to be an attribute of (or at least method of) the SigningProvider?

ishaanam commented on src/psbt.cpp:
```diff
@@ -361,7 +361,7 @@ bool SignPSBTInput(const SigningProvider& provider, PartiallySignedTransaction&
     bool sig_complete;
     if (txdata == nullptr) {
-        sig_complete = ProduceSignature(provider, DUMMY_SIGNATURE_CREATOR, utxo.scriptPubKey, sigdata);
+        sig_complete = ProduceSignature(provider, grind ? DUMMY_SIGNATURE_CREATOR : DUMMY_MAXIMUM_SIGNATURE_CREATOR, utxo.scriptPubKey, sigdata);
```
It seems like the dummy signature produced here isn't used for fee estimation, but for figuring out what signatures that the psbt is missing (for use when analyzing a psbt). It looks like this behavior was introduced in cb40b3abd4514361a024a1e7a1a281da9261261b.

Sjors replied:
@achow101 I could use some hints here.

achow101 replied:
Size estimation occurs in `CreateTransaction`, inside of `SelectCoins` and `CalculateMaximumSignedInputSize`. `FillPSBT` and the functions it calls shouldn't be touched.

ishaanam reviewed (commented):
Concept ACK, but I am a bit confused about the approach so I've left a question below.

Sjors commented:
> A functional test would be nice

Indeed, but I'm not sure how. Can our test suite produce ungrinded signatures?

luke-jr commented:
Worst case, use a hardcoded descriptor and signatures?

Sjors commented:
I changed the approach. `DummySignInput` now checks (the new) `coin_control->m_external_signer` to determine `use_max_sig`. This is much simpler.

luke-jr commented on src/wallet/coincontrol.h:
```diff
@@ -60,6 +60,8 @@ class CCoinControl
     int m_max_depth = DEFAULT_MAX_DEPTH;
     //! SigningProvider that has pubkeys and scripts to do spend size estimation for external inputs
     FlatSigningProvider m_external_provider;
+    //! Wallet inputs are signed by an external signer
+    bool m_external_signer = false;
```
will be*

But I don't think this is a good name for the flag. It's quite possible external signers might support grinding, and will want a way to optimise fees too?

Sjors replied:
I could call it `m_can_grind_r` and default to `true`? And then for external signers we set it to `false` until we have a way (e.g. via HWI) to communicate this capability. That said, I don't know if anyone will implement it, since Taproot doesn't have this problem.

luke-jr commented on src/wallet/spend.cpp:
```diff
@@ -1052,15 +1052,17 @@ util::Result<CreatedTransactionResult> CreateTransaction(
     const auto& txr_ungrouped = *res;
     // try with avoidpartialspends unless it's enabled already
-    if (txr_ungrouped.fee > 0 /* 0 means non-functional fee rate estimation */ && wallet.m_max_aps_fee > -1 && !coin_control.m_avoid_partial_spends) {
+    if (txr_ungrouped.fee > 0 /* 0 means non-functional fee rate estimation */ && wallet.m_max_aps_fee > -1 && !tmp_cc.m_avoid_partial_spends) {
```
No need to change this?

Sjors replied:
Not really, but accessing both the original `coin_control` and `tmp_cc` seems like asking for future regressions.

luke-jr commented:
Also, please don't rebase when there's no reason to. (If you could rebase back, it'd be nice.)

Sjors commented:
> please don't rebase when there's no reason to

Given how much refactoring is (still) going on in the wallet, not rebasing increases the risk of a silent merge conflict. Although in this case back-porting might be very useful, so I'll see if I can move it back to something a bit older.

furszy reviewed (commented):
Instead of adding the `m_external_signer` flag to coin control, what about passing `wallet.IsWalletFlagSet(WALLET_FLAG_EXTERNAL_SIGNER)` to the `DummySignInput` function? (or pass the wallet ref and call `IsWalletFlagSet` internally).

I think that we should keep the coin control object as a container for user modifiable parameters. And not start using it to pass wallet internal flags across functions.

Sjors commented:
@furszy there's three call sites to `DummySignInput`. Two are in `CWallet::DummySignTx` which has direct access to `IsWalletFlagSet()`. The third is in `CalculateMaximumSignedInputSize` (`wallet/spend.cpp`) which only has access to a `SigningProvider`.

I could expand `SigningProvider` with `m_can_grind_r` and `CanGrindR()`. That seems the right place for it, but that might touch a _lot_ of code (haven't tried).

The easier way is to add an `can_grind_r` argument to `CalculateMaximumSignedInputSize`. All it's call sites (outside of tests & bench at least) are in `coin_selection.cpp` and `spend.cpp` and have access to a `CWallet`.

furszy commented:
> I could expand SigningProvider with m_can_grind_r and CanGrindR(). That seems the right place for it, but that might touch a lot of code (haven't tried).

yeah, `SigningProvider` seems to be the right place for it.

> The easier way is to add an can_grind_r argument to CalculateMaximumSignedInputSize. All it's call sites (outside of tests & bench at least) are in coin_selection.cpp and spend.cpp and have access to a CWallet.

Would look like this https://github.com/furszy/bitcoin-core/commit/dccb1205a581bc48e8c3e5230387c72843445587. Not that many changes.

Plus, could also include the unused `CWallet::DummySignTx` removal too: https://github.com/bitcoin/bitcoin/pull/25881/commits/ce81b9a4fa821e92813e62d15d0b1ce5131e0a2e

achow101 commented:
ACK b133ab9b1ee3c824cfc2ca29e3cfc7054377a481

> I could expand `SigningProvider` with `m_can_grind_r` and `CanGrindR()`. That seems the right place for it, but that might touch a _lot_ of code (haven't tried).

I think that would be a layer violation. The `SigningProvider` does not actually do any signing at all, it just provides information that is useful for signing. A `SigningProvider` could be providing information that pertains to both internal and external outputs, in which case the method of signing differs for each input.

Sjors commented:
> I think that would be a layer violation.

I'll leave this PR as-is then. We can always refactor things.

S3RK commented:
Concept ACK. This PR is a clearly an improvement from users' perspective.

Approach wise, I don't like that `CCoinControl` is becoming bloated collection of loosely related parameters. We already have 15 public members there. And from my point of view, ability to grind signatures has nothing to do with coin control. 

I reviewed both this PR and the alternative approach and I like the second option better. You can make it even simpler like this S3RK@8f11705fccbace8534bbf5daf3306465f56c46aa

Sjors commented:
I considered switching to @S3RK's approach of adding extra argument to `DummySignInput` rather than expanding `CCoinControl`. The result is here: https://github.com/Sjors/bitcoin/tree/2023/02/external-signer-feerate-leave-ccoincontrol-alone

The downside of this is approach is that I had to sprinkle `const bool can_grind_r = !wallet->IsWalletFlagSet(WALLET_FLAG_EXTERNAL_SIGNER);` in several different places. This is fine right now, but at some point we might have some external signers that do support grinding. We'd then have to update the check everywhere. So although `CCoinControl` is not really meant to keep track of this, it's the most convenient place.

So I'm keeping the original commit for now.

S3RK commented:
> The downside of this is approach is that I had to sprinkle `const bool can_grind_r = !wallet->IsWalletFlagSet(WALLET_FLAG_EXTERNAL_SIGNER);` in several different places. @furszy's commit does it in three places. This is fine right now, but at some point we might have some external signers that do support grinding. We'd then have to update the check everywhere. 

What if we encapsulate the logic in `Wallet::CanGrindR()`? 

The ability to grind signatures is a characteristic of a wallet IIUC. So that should be pretty future proof.

Sjors commented:
@S3RK done!

furszy commented on src/wallet/spend.cpp:
```diff
@@ -305,7 +305,7 @@ CoinsResult AvailableCoins(const CWallet& wallet,
             std::unique_ptr<SigningProvider> provider = wallet.GetSolvingProvider(output.scriptPubKey);
 
-            int input_bytes = CalculateMaximumSignedInputSize(output, COutPoint(), provider.get(), coinControl);
+            int input_bytes = CalculateMaximumSignedInputSize(output, COutPoint(), provider.get(), wallet.CanGrindR(), coinControl);
```
tiny nit:
could cache this result into a variable outside of the loop (after the `only_safe` variable)

Sjors replied:
Good point, I introduced a `const bool` in both places where it's used in a loop.

Sjors replied:
Though perhaps it's overkill, because `IsWalletFlagSet` is just `m_wallet_flags & flag`.

furszy replied:
yeah, I think that the only place that worth is inside `AvailableCoins`. And not only because of the overhead that introduces (it's one "AND" operation on an atomic variable per wallet UTXO), I was more thinking about the code structure and what we should do if we want to decrease the `cs_wallet` lock contention there.

still, this is still a tiny nit from a person that is thinking about future stuff.

furszy reviewed (approved):
ACK 302f07ee
Looking quite straightforward now.

Sjors committed:
wallet: skip R-value grinding for external signers

When producing a dummy signature for the purpose of estimating the transaction fee, do not assume an external signer performs R-value grinding on the signature.

In particular, this avoids a scenario where the fee rate is 1 sat / vbyte and a transaction with a 72 byte signature is not accepted into our mempool.

This commit also  drops the nullptr default for CCoinControl arguments for functions that it touches. This is because having a boolean argument right next to an optional pointer is error prone.

Co-Authored-By: S3RK <1466284+S3RK@users.noreply.github.com>

furszy reviewed (approved):
ACK 807de2ce

ishaanam reviewed (commented):
utACK 807de2cebdad960c2b52185528ca8960ec694f49

S3RK commented:
ACK 807de2cebdad960c2b52185528ca8960ec694f49

Thanks your patience and addressing the feedback.

achow101 commented:
ACK 807de2cebdad960c2b52185528ca8960ec694f49

Sjors commented on src/wallet/wallet.cpp:
```diff
@@ -1559,7 +1559,7 @@ bool DummySignInput(const SigningProvider& provider, CTxIn &tx_in, const CTxOut
 
     // Use max sig if watch only inputs were used or if this particular input is an external input
     // to ensure a sufficient fee is attained for the requested feerate.
```
Added to a followup PR #27180.
